magit-apply: add literal CR chars if necessary

luismbo opened the pull request:
A tentative fix for #2401, #2403 and #2583.

npostavs commented on lisp/magit-apply.el:
```diff
@@ -86,22 +86,37 @@ With a prefix argument and if necessary, attempt a 3-way merge."
 
+(defun magit-apply--section-file-name (section)
+  (expand-file-name (magit-section-value (magit-section-parent section))
+                    (magit-toplevel)))
```
The relative filename should be enough, no? (or else the git command wouldn't work, I think).

npostavs commented on lisp/magit-apply.el:
```diff
@@ -111,8 +126,7 @@ With a prefix argument and if necessary, attempt a 3-way merge."
                                 (lambda (s)
-                                  (buffer-substring (magit-section-start s)
-                                                    (magit-section-end s)))
+                                  (magit-apply--section-content-for-diff s))
```
You could get rid of the `lambda` too.

npostavs commented on lisp/magit-utils.el:
```diff
@@ -405,6 +405,14 @@ Unless optional argument KEEP-EMPTY-LINES is t, trim all empty lines."
+(defun magit--file-has-crlf-eol-style-p (file)
+  (with-temp-buffer
+      (insert-file-contents-literally file nil 0 500)
+    (and (re-search-forward "\r" nil t) t)))
```
Should be fine to return a generalized boolean so the `and` isn't needed. Also, use `search-forward` since `"\r"` doesn't need to be a regexp.

npostavs commented:
I think you might need to set the `buffer-file-coding-system` explicitly as well. My understanding from #2401 and #2403 is that the patch can sometimes end up with CRs and sometimes not, depending on some setting that I'm unsure of.

Also would be nice to see some more explanatory comments/commit message.

luismbo commented:
I *think* `buffer-file-coding-system` is not relevant here since the patch isn't actually written to a file. I think this approach assumes that `(cdr default-process-coding-system) => utf-8-unix`. I'm not sure what's the right place to codify that requirement. Maybe `magit-run-git-with-input`?

npostavs commented:
> I think buffer-file-coding-system is not relevant here since the patch isn't actually written to a file. I think this approach assumes that (cdr default-process-coding-system) => utf-8-unix.

Ah, that makes sense.

> I'm not sure what's the right place to codify that requirement. Maybe magit-run-git-with-input?

Yeah, that sounds like the right place.

luismbo commented:
@npostavs I've just pushed a new commit that tweaks the code as you suggested and adds a comment, but I'm not feeling particularly inspired. :-)

I don't know how to tweak `magit-run-git-with-input`. (I argue that git *always* expects input to come with LF line endings, so that concern is orthogonal to this pull request. :-P)

npostavs commented:
> adds a comment, but I'm not feeling particularly inspired. :-)

Thanks, now that I see the overview of what's happening, I'm wondering if a better approach is to force the magit buffer to use LF style line endings?

> (I argue that git always expects input to come with LF line endings, so that concern is orthogonal to this pull request. :-P)

Possibly, but my impression is that this is precisely the cause for #2401 and #2403.

luismbo commented:
@npostavs I considered the option of forcing the magit buffer to have LF eol style, but then that would mean that when dealing with projects that use CRLF eol style I would always see the ugly `^M`s at the end of changed lines, and that would be rather inconvenient.

luismbo commented:
Anyway, I'm happy with the current state of this PR if you are too.

npostavs commented on lisp/magit-apply.el:
```diff
@@ -86,22 +86,40 @@ With a prefix argument and if necessary, attempt a 3-way merge."
+;;; to ensure they make their way to the patch file fed to git. The
+;;; remainder of the patch will have LF eol style.
+(defun magit-apply--maybe-fixup-eols (section string)
+  (let ((file (magit-section-value (magit-section-parent section))))
```
I suspect this expression isn't correct (i.e. doesn't return a filename) for all section types, though I haven't got around to testing.

luismbo commented:
I see your point but `magit-apply-patch` makes the same assumption further down in the file.

luismbo commented:
Oh, you're right. The various callers of `magit-apply-patch` pass it different kinds of sections. Sorry about that. Clearly I haven't tested all of those code paths. Will fix.

luismbo commented:
@npostavs should be fixed now! (Thanks for catching that.) I couldn't figure out in what context `magit-apply-diff` is invoked though so I didn't test that one (nor its plural variant).

tarsius commented:
This looks like it is a step in the right direction, but we don't know yet whether this is sufficient or whether this has to be further tweaked in the future. Because of that I think that this should be split up into two commits. The first should add `magit-apply--section-content` and adds/renames local `parent` bindings, i.e. things that on their own could be considered cosmetics. The second commit should be limited to fiddling with line endings. We might have to come back to this a few times and it would be nice if the changes were limited to the issue the commit is trying to solve, not also related refactorings and cleanup.

luismbo commented:
@tarsius done. The exercise you suggested helped me further simplify the code. (Nice!) BTW, rewriting the initial commit using magit is a joy. ;-)

luismbo commented:
Added a variable as suggested. (Naming is hard, writing documentation is hard... Uf.)

Regarding `core.autocrlf` et al., my personal experience tells me that doing that sort of automatic CRLF <-> LF conversion in source code is a recipe for disaster.

tarsius commented:
> (Naming is hard, writing documentation is hard... Uf.)

Yeah :-)

What about:

```
(defcustom magit-preserve-crlf-line-edings t
  "Whether to try harder to preserve CRLF line ending type.

Magit buffers may contain diffs for more than one file and these
files might use different line ending types, but Emacs normalizes
file endings in a buffer to just one type, usually LF.

This isn't an issue when staging (or otherwise applying) complete
files, because the buffer contents are not used in that case.
But when applying hunks, then a patch has to be build from parts
of the buffer contents.

If this option is non-nil (the default), then Magit peeks the
file that is being changed to determine whether its line ending
style is CRLF.  If so, then CR characters are inserted into the
patch that is being sent to git.

Also see the Git variables `core.eol', `core.safecrlf', and
`core.autocrlf' for an alternative approach."
  :package-version '(magit . "2.6.0")
  :group 'magit
  :type 'boolean)
```

Also could you please mention this in the release notes.

@npostavs Minor issues aside, do you think this is good to go? (I would like to create a new release after merging the two remaining pull requests. /cc @kyleam)

npostavs commented:
> Minor issues aside, do you think this is good to go?

I think so, but I haven't actually tested it.

npostavs commented on lisp/magit-utils.el:
```diff
@@ -405,6 +405,14 @@ Unless optional argument KEEP-EMPTY-LINES is t, trim all empty lines."
+
+(defun magit--file-has-crlf-eol-style-p (file)
+  (with-temp-buffer
+      (insert-file-contents-literally file nil 0 500)
```
This lines is mis-indented.

luismbo replied:
Oops. Fixed.

luismbo commented:
Ugh, sorry I missed the `defcustom` arguments. I've tweaked the documentation string as follows:

```diff
modified   lisp/magit-apply.el
@@ -72,11 +72,11 @@
 
 Magit buffers may contain diffs for more than one file and these
 files might use different line ending types, but Emacs normalizes
-file endings in a buffer to just one type, usually LF.
+file endings in a buffer to a single newline character.
 
 This isn't an issue when staging (or otherwise applying) complete
 files, because the buffer contents are not used in that case.
-But when applying hunks, then a patch has to be build from parts
+But when applying hunks, then a patch has to be built from parts
 of the buffer contents.
 
 If this option is non-nil (the default), then Magit peeks the
```

Added the following entry to the release notes in the fixes section:

```
* Staging hunks/regions belonging to files with CRLF line endings on
  Windows (or, to be precise, when the Magit status buffer uses CRLF
  end-of-line style) ended up erroneously staging changes with LF line
  endings. Magit now tries harder to not mangle line endings. A new
  variable `magit-preserve-crlf-line-endings' has been added to allow
  disabling the new behavior, should it cause unforseen troubles.
```

luismbo commented:
Forgot to update the variable name in the commit message. Fixed in d46419e.

tarsius commented:
I just tried reproducing the issue on Linux, using both Emacs 24.5 and 25.0.50, but couldn't.

In my case each line (added, removed, and context) in the CRLF file ends with `^M` in the diff in the status buffer. Git only shows the `^M` for added and remove lines, but I think that is an optimization. If I stage a hunk in Magit the staged changes still end with `^M` and when I commit the eol style is also preserved, which I verified using both Magit and Git directly.

The coding system in the buffer visiting the file is `undecided-dos` and in the Magit buffer it is `utf-8-unix`. Does that match what you see?

In the past users have reported that they see the `^M` and we might look into *hiding* them, but I think this is correct. Magit buffers should always use `utf-8-unix` - since they might contain part from different files we should use a reasonable default.

However Magit never actually enforces that, it just works because that's what Emacs does by default. I am not sure though whether that is always the case, maybe on Windows it doesn't, which would explain why you don't see the `^M`, or maybe this depends on more than just the OS. @npostavs could you too please try on Windows?

If we come to the conclusion that this always happens on Windows, then we should set the option value based on the `system-type`. Otherwise it should be off by default, I think.

tarsius commented on lisp/magit-apply.el:
```diff
@@ -86,22 +109,43 @@ With a prefix argument and if necessary, attempt a 3-way merge."
+;;; endings. To recover that information, we check if the file to be
+;;; patched has CRLF lines, in which case we insert literal CR chars
+;;; to ensure they make their way to the patch file fed to git. The
+;;; remainder of the patch will have LF eol style.
```
Let's drop that completely and concentrate on the option's doc-string (which the reader of the below code can easily find because the option is used). Also comments should begin with just two semicolons. Using more turns these lines into seven section headings when using `outline-minor-mode`.

luismbo replied:
Dropped in a23f473. Using three semicolons for top-level comments is a CL convention, sorry about that.

tarsius commented on lisp/magit-utils.el:
```diff
@@ -405,6 +405,14 @@ Unless optional argument KEEP-EMPTY-LINES is t, trim all empty lines."
+(defun magit--file-has-crlf-eol-style-p (file)
+  (with-temp-buffer
+    (insert-file-contents-literally file nil 0 500)
+    (search-forward "\r" nil t)))
```
Instead of doing that, have you considered using `(coding-system-eol-type buffer-file-coding-system)` here too?

#2584

luismbo replied:
I did consider that, but my emacs-fu was not strong enough to figure out how to quietly open the file using `find-file`. `with-temp-buffer` doesn't work for that because it will have some default `buffer-file-coding-system` to start with.

tarsius replied:
> figure out how to quietly open the file using `find-file`.

You can use `find-file-noselect`.

> `with-temp-buffer` doesn't work for that because it will have some default `buffer-file-coding-system` to start with.

Ah, didn't realize that. But using `find-file-noselect` would be problematic because it runs a lot of stuff we don't want and if the file is already being visited in a buffer... So It's probably better to stick with what you have done. (But 500 seems a bit low.)

luismbo replied:
How about something like this?

```elisp
(defun magit--file-has-crlf-eol-style-p (file)
  (with-temp-buffer
    (cl-loop with start = 0
             for (_ bytes-read) = (insert-file-contents-literally file nil start 500)
             until (zerop bytes-read)
             do (incf start bytes-read)
             when (re-search-forward "(\r)?\n" nil t)
             return (string= (match-string 1) "\r"))))
```

(It doesn't work because the regex is botched, but you get the general idea.)

tarsius replied:
Couldn't that mean that we end up in the middle of some character? Probably just leave it as is.

luismbo replied:
I think `re-search-forward` doesn't move the point if it doesn't find anything, so it'd be okay, but I agree it's probably better to leave the current, simpler, version as is.

luismbo commented:
> I just tried reproducing the issue on Linux, using both Emacs 24.5 and 25.0.50, but couldn't.

If this is a surprise for you, then I have failed to convey what is
going on. (Perhaps the `magit-preserve-crlf-line-endings` docstring
needs further tweaking then!) The problem only occurs when Magit's
status buffer has CRLF encoding (and that should only occur on
Windows).

There is no problem when Magit status buffer has LF encoding because
in that case a change coming from file with lines ending with two
bytes (CR and LF) are translated into two character in the buffer:
`?\r` and `?\n`. However, when Magit's status buffer has CRLF
encoding, those same two bytes (CR and LF) are translated into a
single a character in the status buffer: `?\n`.

Does it make more sense now?

tarsius commented:
> Does it make more sense now?

Yes! (But you explained that well enough before, it's just that I find it very unintuitive and appear to have over read the important parts. Why would it be an issue if both the buffer and the file used the *same* eol style?! But I get it now.)

> The problem only occurs when Magit's status buffer has CRLF encoding

Maybe we should figure out how to use LF encoding in Magit buffers? If I understand everything correctly that shouldn't cause any issues - except of the `^M`s everywhere. Which we could hide, but that might be *more* expensive than what you are suggesting, so...

> (and that should only occur on Windows).

In that case the option's default value should be set using `(and (memq system-type '(ms-dos windows-nt)) t)` (but not `cygwin`, right?). That way only those you might benefit have to pay the price.

I finally start realizing that this probably is the *best* approach :-) By the way, thanks for hanging in there! That's the sort of issue that only gets resolved if an affected user invests the necessary time (because my understanding is spotty and I also have no way of testing.)

> Perhaps the `magit-preserve-crlf-line-endings` docstring needs further tweaking then!

Yes, please incorporate the information from your last message.

tarsius commented on lisp/magit-apply.el:
```diff
@@ -86,22 +109,36 @@ With a prefix argument and if necessary, attempt a 3-way merge."
       (`(,_  files) (magit-apply-diffs  it args)))))
 
+(defun magit-apply--maybe-fixup-eols (section string)
+  (let ((file (magit-section-value section)))
```
Okay, now I start nitpicking (and it's totally a matter of personal preference), but since the value is only used once, I don't see a reason to use a variable here.

luismbo replied:
> Okay, now I start nitpicking (and it's totally a matter of personal preference), but since the value is only used once, I don't see a reason to use a variable here.


I don't have strong feelings either way. Changed as suggested in 5999c20.

tarsius commented on lisp/magit-apply.el:
```diff
@@ -86,22 +109,36 @@ With a prefix argument and if necessary, attempt a 3-way merge."
+    (if (and magit-preserve-crlf-line-endings
+             (magit--buffer-has-crlf-eol-style-p)
+             (magit--file-has-crlf-eol-style-p file))
+        (replace-regexp-in-string "\n" "\r\n" string)
```
It might be worth investigating whether this can be done more efficiently using `line-end-position`. But that would have to happen in `magit-apply--section-contents` and maybe `conat`enating many times is *less* efficient than just using `buffer-substring` and `replace-regexp-in-string`. And it would get a lot more complicated.

tarsius replied:
In other words, let's just leave this as is too.

tarsius commented on lisp/magit-apply.el:
```diff
@@ -86,22 +109,36 @@ With a prefix argument and if necessary, attempt a 3-way merge."
+   (pcase (magit-section-type section)
+     (`file section)
+     (`hunk (magit-section-parent section)))
+   (buffer-substring (magit-section-start section)
```
But we should use `buffer-substring-no-properties` here. And since that isn't really related to the other changes, it should be done using a separate commit in between the other two.

luismbo replied:
On Sat, Mar 12, 2016 at 10:45 AM, Jonas Bernoulli
<notifications@github.com> wrote:
> But we should use buffer-substring-no-properties here. And since that isn't
> really related to the other changes, it should be done using a separate
> commit in between the other two.

Done in 8450a3e.

luismbo commented:
> In that case the option's default value should be set using (and (memq system-type '(ms-dos windows-nt)) t) (but not cygwin, right?). That way only those you might benefit have to pay the price.

The price of peeking the file is payed if, and only if,
`magit--buffer-has-crlf-eol-style-p` returns t. That will only happen
on Windows. I think that check is more robust than a hard-coded list
of `system-type`s.


> > Perhaps the magit-preserve-crlf-line-endings docstring needs further tweaking then!
>
> Yes, please incorporate the information from your last message.

Done in 5999c20.

tarsius commented:
> I think that check is more robust than a hard-coded list of `system-type`s.

True. But that (me not realizing that we don't do anything expensive if we don't have to) reminds me that initially I thought that `magit--{buffer,file}-has-crlf-eol-style-p` should just be folded into `magit-apply--maybe-fixup-eols`. Since these functions are tiny and only used here I don't see much benefit in keeping them around. And if we stick the code directly into the fixup function, then one can immediately see that determining the eol style for the buffer is cheap while the same is more costly for the file. (I would have realized that anyway instead of making the previous suggestion, but still.)

luismbo commented:
My reasoning there was that

```elisp
(defun magit-apply--maybe-fixup-eols (section string)
  (if (and magit-preserve-crlf-line-endings
           (magit--buffer-has-crlf-eol-style-p)
           (magit--file-has-crlf-eol-style-p (magit-section-value section)))
      (replace-regexp-in-string "\n" "\r\n" string)
      string))
```

is less obscure than

```elisp
(defun magit-apply--maybe-fixup-eols (section string)
  (if (and magit-preserve-crlf-line-endings
           (eql 1 (coding-system-eol-type buffer-file-coding-system))
           (with-temp-buffer
             (insert-file-contents-literally (magit-section-value section)
                                             nil 0 500)
             (search-forward "\r" nil t)))
      (replace-regexp-in-string "\n" "\r\n" string)
      string))
```

Perhaps adding some comments helps, but when in doubt I try to make code self-readable rather than adding comments.

```elisp
(defun magit-apply--maybe-fixup-eols (section string)
  (if (and magit-preserve-crlf-line-endings
           ;; buffer has CRLF eol type
           (eql 1 (coding-system-eol-type buffer-file-coding-system))
           ;; file has CRLF eol type, too
           (with-temp-buffer
             (insert-file-contents-literally (magit-section-value section)
                                             nil 0 500)
             (search-forward "\r" nil t)))
      (replace-regexp-in-string "\n" "\r\n" string)
      string))
```

I don't have hard feelings for either of the three versions. Let me know which one you prefer.

npostavs commented:
(I don't have access to a Windows system with magit right at the moment)

>> I just tried reproducing the issue on Linux, using both Emacs 24.5 and 25.0.50, but couldn't.
>
> If this is a surprise for you, then I have failed to convey what is going on. (Perhaps the `magit-preserve-crlf-line-endings` docstring needs further tweaking then!) The problem only occurs when Magit's status buffer has CRLF encoding (and that should only occur on Windows).

(Using my GNU/Linux box) I tried setting `buffer-file-coding-system` to `utf-8-dos`, but I still see the `^M`s in the status buffer.

luismbo commented:
@npostavs `buffer-file-coding-system` needs to be set before the buffer is constructed. Perhaps setting `default-buffer-file-coding-system` does the trick.

npostavs commented:
Hmm, I tried `(setq-default buffer-file-coding-system 'utf-8-dos)` and `(setq default-buffer-file-coding-system 'utf-8-dos)` but no difference.

tarsius commented:
Go with the third. I'll merge once that is done.

luismbo commented:
@npostavs gosh. It's actually not `buffer-file-coding-system` that's at fault here. Try this instead:

```
(setq default-process-coding-system '(utf-8-dos . utf-8-unix))
```

I'll have to think a little bit about the implications of this.

luismbo commented:
Alright. So, the general idea of this pull request is still correct, except the information lossage is not due to `buffer-file-coding-system` but due to `process-coding-system` used when grabbing the diffs from git. I see two options to proceed:

  1. Tweak this pull request to reflect this new insight (i.e. check `(car default-process-coding-system)` instead of `buffer-file-coding-system`) and live with the fact that `default-process-coding-system` might change between grabbing the diffs and `magit-apply`ing. (Fairly unlikely, but still...)
  2. Change `magit-process` to ensure that `process-coding-system` has an LF eol style and (optionally) hide them in the Magit buffer using something like
```elisp
(setq buffer-display-table (make-display-table))
(aset buffer-display-table ?\^M [])
```

(I might need some help if we want to pursue the 2nd option, both for tweaking `magit-process` and for figuring out the best way to hide the `^M` chars.)

What do you guys think?

npostavs commented on lisp/magit-apply.el:
```diff
@@ -86,22 +118,40 @@ With a prefix argument and if necessary, attempt a 3-way merge."
+
+(defun magit-apply--section-content (section)
+  (magit-apply--maybe-fix-up-eols
+   (case (magit-section-type section)
```
should be `cl-case`

luismbo replied:
Ups. Fixed in bceef47 to use `pcase` again. (Noticed that `pcase` is used extensively throughout magit, never `cl-case`.)

npostavs commented:
>  Change `magit-process` to ensure that `process-coding-system` has an LF eol style and (optionally) hide them in the Magit buffer

I like this method a bit better, it seems less likely to go wrong.

luismbo commented:
I've pushed bceef47, which implements the 1st option, just in case. I'll start exploring the 2nd option now.

luismbo commented:
Should I open another pull request to pursue that option?

tarsius commented:
> Should I open another pull request to pursue that option?

Whatever works best for you. You could also just use a temporary commit, to be later either squashed or discarded.

luismbo commented:
Closing in favor of #2587.
